chain/vm/fvm.go

```diff
@@ -303,6 +308,111 @@ func NewFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
 	}, nil
 }
 
+func NewDebugFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
+	state, err := state.LoadStateTree(cbor.NewCborStore(opts.Bstore), opts.StateBase)
+	if err != nil {
+		return nil, err
+	}
+
+	circToReport, err := opts.CircSupplyCalc(ctx, opts.Epoch, state)
+	if err != nil {
+		return nil, err
+	}
+
+	baseBstore := opts.Bstore
+	overlayBstore := blockstore.NewMemorySync()
+	cborStore := cbor.NewCborStore(overlayBstore)
+	vmBstore := blockstore.NewTieredBstore(overlayBstore, baseBstore)
+
+	fvmopts := &ffi.FVMOpts{
+		FVMVersion: 0,
+		Externs: &FvmExtern{
+			Rand:       opts.Rand,
+			Blockstore: vmBstore,
+			lbState:    opts.LookbackState,
+			base:       opts.StateBase,
+			epoch:      opts.Epoch,
+		},
+		Epoch:          opts.Epoch,
+		BaseFee:        opts.BaseFee,
+		BaseCircSupply: circToReport,
+		NetworkVersion: opts.NetworkVersion,
+		StateBase:      opts.StateBase,
+		Tracing:        EnableDetailedTracing,
+		Debug:          true,
+	}
+
+	putMapping := func(ar map[cid.Cid]cid.Cid) (cid.Cid, error) {
+		var mapping xMapping
+
+		mapping.redirects = make([]xRedirect, 0, len(ar))
+		for from, to := range ar {
+			mapping.redirects = append(mapping.redirects, xRedirect{from: from, to: to})
+		}
+		sort.Slice(mapping.redirects, func(i, j int) bool {
+			return bytes.Compare(mapping.redirects[i].from.Bytes(), mapping.redirects[j].from.Bytes()) < 0
+		})
+
+		// Passing this as a pointer of structs has proven to be an enormous PiTA; hence this code.
+		mappingCid, err := cborStore.Put(context.TODO(), &mapping)
+		if err != nil {
+			return cid.Undef, err
+		}
+
+		return mappingCid, nil
+	}
+
+	av, err := actors.VersionForNetwork(opts.NetworkVersion)
+	if err != nil {
+		return nil, xerrors.Errorf("error determining actors version for network version %d: %w", opts.NetworkVersion, err)
+	}
+
+	switch av {
+	case actors.Version8:
+		if bundlePath := os.Getenv("LOTUS_FVM_DEBUG_BUNDLE_V8"); bundlePath != "" {
+			mfCid, err := bundle.LoadBundleFromFile(ctx, overlayBstore, bundlePath)
+			if err != nil {
+				return nil, err
+			}
+
+			mf, err := actors.LoadManifestData(ctx, mfCid, adt.WrapStore(ctx, cborStore))
+			if err != nil {
+				return nil, err
+			}
+
+			// create actor redirect mapping
+			actorRedirect := make(map[cid.Cid]cid.Cid)
+			for _, e := range mf.Entries {
+				from, ok := actors.GetActorCodeID(actors.Version8, e.Name)
+				if !ok {
+					log.Warnf("unknown actor %s", e.Name)
+					continue
+				}
+
+				actorRedirect[from] = e.Code
+			}
+
+			if len(actorRedirect) > 0 {
+				mappingCid, err := putMapping(actorRedirect)
+				if err != nil {
+					return nil, xerrors.Errorf("error writing redirect mapping: %w", err)
+				}
+				fvmopts.ActorRedirect = mappingCid
```

Comment:
nit: can we put this in a separate helper function?

Reply:
yuss, good call